ocaml/xenopsd/lib/xenops_server.ml

```diff
@@ -1312,6 +1334,12 @@ let vgpu_receiver_sync : (Vm.id, vgpu_fd_info) Hashtbl.t = Hashtbl.create 10
 
 let vgpu_receiver_sync_m = Mutex.create ()
 
+type mem_fd_info = {mem_fd: Unix.file_descr; mem_channel: unit Event.channel}
+
+let mem_receiver_sync : (Vm.id, mem_fd_info) Hashtbl.t = Hashtbl.create 10
+
+let mem_receiver_sync_m = Mutex.create ()
```

Comment:
It looks like modification is done using `with_lock` on this lock and the hashtable, but read is done without using the lock. I think that might still be unsafe: the hashtable might get modified as we read it, so the lock should be used in both cases, ideally the raw `mem_receiver_sync` should not be exposed and all accesses should be done through some wrapper functions that always take the lock.

Otherwise this may be a bit fragile and rely on an implementation detail: OCaml can currently only switch threads when a C stub call is made that relinquishes the runtime lock, OR when the called function allocates (and then the signal queued by the periodic ~50ms timer action gets processed that switches threads). I'm not sure we can rely on Hashtbl.find_opt never allocating (apart from the final `Some data`), and OCaml 5.0 would change the "only one OCaml thread may run at one time" if we start using the new Domains feature causing the hashtable to get mutated as we're searching it which might cause us not to find an entry that is otherwise there.

We probably already have this problem with various other global hashtables used in xenopsd, we should probably review its thread safety, regardless of multicore multiple threads are already being used in xenopsd, and it probably happens to work because *currently* Hashtbl.find_opt doesn't allocate during a search.

Reply:
I have rebased the code and added locking around reading the hash table.

Reply:
I think you forgot one: see below.